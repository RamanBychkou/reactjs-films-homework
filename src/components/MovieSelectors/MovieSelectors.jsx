import React from 'react';
import shortid from 'shortid';
import PropTypes from 'prop-types';
import selectGenre from '../../utils/selectGenre';
import style from '../../pages/MovieListContainer/MovieListContainer.scss';

export const createGenreList = genres => genres.map((c) => {
  const key = shortid.generate();
  return <option key={key}>{c.name}</option>;
});

class MovieSelectors extends React.Component {
  shooseTypeView = (e) => {
    const { setTypeView } = this.props;
    const typeName = e.target.id;
    if (typeName === 'cards') {
      return setTypeView('cards');
    }
    return setTypeView('lines');
  };

  showTrends = (e) => {
    const query = e.target.textContent;
    const { fetchListMovies, error, clearError } = this.props;
    if (error !== undefined && error !== '') {
      clearError('');
    }
    return fetchListMovies(query);
  };

  render() {
    const { genres } = this.props;
    return (
      <div className={style.moviesSelector}>
        <nav>
          <ul>
            <li>
              <button type="button" href="#" onClick={this.showTrends}>
                Trending
              </button>
            </li>
            <li>
              <button type="button" href="#" onClick={this.showTrends}>
                Top Rated
              </button>
            </li>
            <li>
              <button type="button" href="#" onClick={this.showTrends}>
                Coming Soon
              </button>
            </li>
            <li>
              <select name="genre" id="" onChange={selectGenre.bind(this, this.props)}>
                <option value="">Genre</option>
                {createGenreList(genres)}
              </select>
            </li>
          </ul>
        </nav>
        <div className={style.typeView}>
          <button type="button" id="cards" onClick={this.shooseTypeView}>
            &#9871;
          </button>
          <button type="button" id="lines" onClick={this.shooseTypeView}>
            &#9868;
          </button>
        </div>
      </div>
    );
  }
}

MovieSelectors.defaultProps = {
  genres: [],
  error: '',
};

MovieSelectors.propTypes = {
  genres: PropTypes.arrayOf(PropTypes.any),
  setTypeView: PropTypes.func.isRequired,
  fetchListMovies: PropTypes.func.isRequired,
  clearError: PropTypes.func.isRequired,
  error: PropTypes.string,
};

export default MovieSelectors;
