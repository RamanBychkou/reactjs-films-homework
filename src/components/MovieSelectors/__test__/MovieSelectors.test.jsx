import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils, { act } from 'react-dom/test-utils';
import { MemoryRouter as Router } from 'react-router-dom';
import ReactTestRender from 'react-test-renderer';
import MovieSelectors from '../index';

describe('MovieSelectors', () => {
  const genres = [
    { id: 28, name: 'Action' },
    { id: 12, name: 'Adventure' },
    { id: 16, name: 'Animation' },
    { id: 35, name: 'Comedy' },
    { id: 80, name: 'Crime' },
    { id: 99, name: 'Documentary' },
    { id: 18, name: 'Drama' },
    { id: 10751, name: 'Family' },
    { id: 14, name: 'Fantasy' },
    { id: 36, name: 'History' },
    { id: 27, name: 'Horror' },
    { id: 10402, name: 'Music' },
    { id: 9648, name: 'Mystery' },
    { id: 10749, name: 'Romance' },
    { id: 878, name: 'Science Fiction' },
    { id: 10770, name: 'TV Movie' },
    { id: 53, name: 'Thriller' },
    { id: 10752, name: 'War' },
    { id: 37, name: 'Western' },
  ];
  const setTypeView = jest.fn();
  const clearError = jest.fn(() => 1);
  it('MovieSelectors: renders correctly', () => {
    const result = ReactTestRender.create(
      <Router>
        <MovieSelectors genres={genres} setTypeView={setTypeView} clearError={clearError} />
      </Router>,
    );
    expect(result).toMatchSnapshot();
  });

  describe('MovieSelectors: events click ', () => {
    const fetchListMovies = jest.fn();
    const test = jest.fn(value => value);
    const props = {
      fetchListMovies,
      error: 'true',
      clearError,
      setTypeView: test,
      history: { push: jest.fn() },
    };
    let container;

    beforeEach(() => {
      container = document.createElement('div');
      container.id = 'modalRoot';
      document.body.appendChild(container);
      act(() => {
        ReactDOM.render(
          <Router>
            <MovieSelectors {...props} />
          </Router>,
          document.querySelector('#modalRoot'),
        );
      });
    });

    // it('selectGenre: error true should BE - Call clearError', () => {
    //   props.genres = [{ name: 'Drame', id: 35 }];
    //   const btn = React.createElement(
    //     'button',
    //     {
    //       id: 'test',
    //       onClick: selectGenre.bind(null, props),
    //       type: 'button',
    //     },
    //     'Drama',
    //   );
    //   jest.spyOn(props, 'clearError');
    //   act(() => {
    //     ReactDOM.render(btn, container);
    //   });
    //   const node = container.querySelector('#test');
    //   ReactTestUtils.Simulate.click(node);
    //   expect(clearError).toHaveBeenCalled();
    // });
    // it('showTrends: error true should BE - call clearError', () => {
    //   jest.spyOn(props, 'clearError');
    //   const node = document.querySelector('button');
    //   ReactTestUtils.Simulate.click(node, { target: { textContent: 'Coming Soon' } });
    //   expect(clearError).toHaveBeenCalled();
    // });

    it('shooseTypeView: shooseTypeView - call action', () => {
      jest.spyOn(props, 'setTypeView');
      const node1 = document.querySelector('#cards');
      const node2 = document.querySelector('#lines');
      ReactTestUtils.Simulate.click(node1);
      ReactTestUtils.Simulate.click(node2);
      expect(props.setTypeView).toHaveBeenCalledTimes(2);
    });
  });
  describe('showTrends: without error', () => {
    test('showTrends: without error - clearError not called with undefined', () => {
      let container;

      beforeEach(() => {
        const props = {
          history: { push: jest.fn() },
          clearError,
          setTypeView,
        };
        container = document.createElement('div');
        container.id = 'modalRoot';
        document.body.appendChild(container);
        act(() => {
          ReactDOM.render(
            <Router>
              <MovieSelectors {...props} />
            </Router>,
            document.querySelector('#modalRoot'),
          );
        });
        jest.spyOn(props, 'clearError');
        const node = document.querySelector('button');
        ReactTestUtils.Simulate.click(node);
        expect(clearError).toHaveBeenCalledTimes(0);
      });
    });
    test('showTrends: without error - clearError not called with ""', () => {
      let container;

      beforeEach(() => {
        const props = {
          error: 'errror',
          clearError,
        };
        container = document.createElement('div');
        container.id = 'modalRoot';
        document.body.appendChild(container);
        act(() => {
          ReactDOM.render(
            <Router>
              <MovieSelectors {...props} />
            </Router>,
            document.querySelector('#modalRoot'),
          );
        });
        jest.spyOn(props, 'clearError');
        const node = document.querySelector('button');
        ReactTestUtils.Simulate.click(node);
        expect(clearError).toHaveBeenCalledTimes(0);
      });
    });
  });
});
