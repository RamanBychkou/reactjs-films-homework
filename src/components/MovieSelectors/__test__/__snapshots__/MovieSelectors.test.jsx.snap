// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieSelectors MovieSelectors: renders correctly 1`] = `
<div
  className="moviesSelector"
>
  <nav>
    <ul>
      <li>
        <button
          href="#"
          onClick={[Function]}
          type="button"
        >
          Trending
        </button>
      </li>
      <li>
        <button
          href="#"
          onClick={[Function]}
          type="button"
        >
          Top Rated
        </button>
      </li>
      <li>
        <button
          href="#"
          onClick={[Function]}
          type="button"
        >
          Coming Soon
        </button>
      </li>
      <li>
        <select
          id=""
          name="genre"
          onChange={[Function]}
        >
          <option
            value=""
          >
            Genre
          </option>
          <option>
            Action
          </option>
          <option>
            Adventure
          </option>
          <option>
            Animation
          </option>
          <option>
            Comedy
          </option>
          <option>
            Crime
          </option>
          <option>
            Documentary
          </option>
          <option>
            Drama
          </option>
          <option>
            Family
          </option>
          <option>
            Fantasy
          </option>
          <option>
            History
          </option>
          <option>
            Horror
          </option>
          <option>
            Music
          </option>
          <option>
            Mystery
          </option>
          <option>
            Romance
          </option>
          <option>
            Science Fiction
          </option>
          <option>
            TV Movie
          </option>
          <option>
            Thriller
          </option>
          <option>
            War
          </option>
          <option>
            Western
          </option>
        </select>
      </li>
    </ul>
  </nav>
  <div
    className="typeView"
  >
    <button
      id="cards"
      onClick={[Function]}
      type="button"
    >
      ⚏
    </button>
    <button
      id="lines"
      onClick={[Function]}
      type="button"
    >
      ⚌
    </button>
  </div>
</div>
`;
